paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/catalog/functions/BucketFunction.java

```diff
@@ -0,0 +1,724 @@
+import org.apache.spark.sql.types.TimestampType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/** A Spark function implementation for the Paimon bucket transform. */
+public class BucketFunction implements UnboundFunction {
+    private static final int NUM_BUCKETS_ORDINAL = 0;
+    private static final int SPARK_TIMESTAMP_PRECISION = 6;
+
+    private static final Map<String, Class<? extends BucketGeneric>> BUCKET_FUNCTIONS;
+
+    static {
+        ImmutableMap.Builder<String, Class<? extends BucketGeneric>> builder =
+                ImmutableMap.builder();
+        builder.put("BucketBoolean", BucketBoolean.class);
+        builder.put("BucketByte", BucketByte.class);
+        builder.put("BucketShort", BucketShort.class);
+        builder.put("BucketInteger", BucketInteger.class);
+        builder.put("BucketLong", BucketLong.class);
+        builder.put("BucketFloat", BucketFloat.class);
+        builder.put("BucketDouble", BucketDouble.class);
+        builder.put("BucketString", BucketString.class);
+        builder.put("BucketDecimal", BucketDecimal.class);
+        builder.put("BucketTimestamp", BucketTimestamp.class);
+        builder.put("BucketBinary", BucketBinary.class);
+
+        // Joint bucket fields of common types
+        builder.put("BucketIntegerInteger", BucketIntegerInteger.class);
+        builder.put("BucketIntegerLong", BucketIntegerLong.class);
+        builder.put("BucketIntegerString", BucketIntegerString.class);
+        builder.put("BucketLongInteger", BucketLongInteger.class);
+        builder.put("BucketLongLong", BucketLongLong.class);
+        builder.put("BucketLongString", BucketLongString.class);
+        builder.put("BucketStringInteger", BucketStringInteger.class);
+        builder.put("BucketStringLong", BucketStringLong.class);
+        builder.put("BucketStringString", BucketStringString.class);
+
+        BUCKET_FUNCTIONS = builder.build();
+    }
+
+    public static boolean supportsTable(FileStoreTable table) {
+        if (table.bucketMode() != BucketMode.HASH_FIXED) {
+            return false;
+        }
+
+        return table.schema().logicalBucketKeyType().getFieldTypes().stream()
+                .allMatch(BucketFunction::supportsType);
+    }
+
+    private static boolean supportsType(org.apache.paimon.types.DataType type) {
+        if (type instanceof ArrayType
+                || type instanceof MapType
+                || type instanceof RowType
+                || type instanceof VariantType) {
+            return false;
+        }
+
+        if (type instanceof org.apache.paimon.types.TimestampType) {
+            return ((org.apache.paimon.types.TimestampType) type).getPrecision()
+                    == SPARK_TIMESTAMP_PRECISION;
```

Comment:
The Spark Timestamp type is in microseconds and does not allow custom precision. When binding the Paimon bucket function, Paimon’s type information is not available. Therefore, support is only possible when the Paimon timestamp precision is equal to the Spark timestamp precision.